src/llmq/quorums_chainlocks.cpp

```diff
@@ -0,0 +1,350 @@
+        }
+        if (it->second.blockHash != recoveredSig.msgHash) {
+            // does not match out vote
+            return;
+        }
+
+        it->second.sig = recoveredSig.sig;
+        clsig = it->second;
+    }
+    ProcessNewChainLock(-1, clsig, ::SerializeHash(clsig));
+}
+
+void CChainLocksHandler::ScheduleInvalidateBlock(const CBlockIndex* pindex)
+{
+    // Calls to InvalidateBlock and ActivateBestChain might result in re-invocation of the UpdatedBlockTip and other
+    // signals, so we can't directly call it from signal handlers. We solve this by doing the call from the scheduler
+
+    scheduler->scheduleFromNow([this, pindex]() {
+        DoInvalidateBlock(pindex, true);
+    }, 0);
+}
+
+void CChainLocksHandler::DoInvalidateBlock(const CBlockIndex* pindex, bool activateBestChain)
+{
+    auto& params = Params();
+
+    {
+        LOCK(cs_main);
+
+        // get the non-const pointer
+        CBlockIndex* pindex2 = mapBlockIndex[pindex->GetBlockHash()];
+
+        CValidationState state;
+        if (!InvalidateBlock(state, params, pindex2)) {
+            LogPrintf("CChainLocksHandler::UpdatedBlockTip -- InvalidateBlock failed: %s\n", FormatStateMessage(state));
+            // This should not have happened and we are in a state were it's not safe to continue anymore
+            assert(false);
+            return;
+        }
+    }
+
+    CValidationState state;
+    if (activateBestChain && !ActivateBestChain(state, params)) {
+        LogPrintf("CChainLocksHandler::UpdatedBlockTip -- ActivateBestChain failed: %s\n", FormatStateMessage(state));
+        // This should not have happened and we are in a state were it's not safe to continue anymore
+        assert(false);
+        return;
+    }
+}
+
+bool CChainLocksHandler::HasChainLock(const CBlockIndex* pindexPrev, const uint256& newBlockHash)
+{
+    if (!sporkManager.IsSporkActive(SPORK_19_CHAINLOCKS_ENABLED)) {
+        return false;
+    }
+
+    LOCK2(cs_main, cs);
+    return InternalHasChainLock(pindexPrev, newBlockHash);
+}
+
+bool CChainLocksHandler::InternalHasChainLock(const CBlockIndex* pindexPrev, const uint256& newBlockHash)
+{
+    AssertLockHeld(cs_main);
+    AssertLockHeld(cs);
+
+    auto it = chainLocksByPrev.find(pindexPrev->GetBlockHash());
+    if (it == chainLocksByPrev.end()) {
+        return false;
+    }
+
+    if (it->second->blockHash != newBlockHash) {
+        // another block was locked at this height
+        return false;
+    }
+
+    return true;
+}
+
+bool CChainLocksHandler::HasChildChainLock(const CBlockIndex* pindex)
+{
+    if (!sporkManager.IsSporkActive(SPORK_19_CHAINLOCKS_ENABLED)) {
+        return false;
+    }
+
+    LOCK2(cs_main, cs);
+
+    // TODO carry the locked state in CBlockIndex instead of searching for last ChainLock?
+
+    while (pindex) {
+        if (InternalHasChainLock(pindex->pprev, pindex->GetBlockHash())) {
+            return true;
+        }
+        pindex = chainActive.Next(pindex);
```

Comment:
This can be pretty expensive for early blocks. Maybe introduce a limit it for how many blocks we are going to keep chainlocks (4032 i.e. 1 week) and if pindex is twice this deep from the tip just bail out (because we know we do not store that much)?

Reply:
Also, why would we want to do this at all? Why don't we just show `true` for blocks/headers we actually have chainlocks for?

Reply:
The reason to show `true` for blocks below a CL is that a CL implicitly locks all it's parents as well. I wanted to make it as easy as possible for users of the RPCs, as otherwise they would have to implement this logic.

I'll try to come up with a logic that is not so expensive.

Reply:
I like d25d35e, this looks much much much more efficient :)
Re storing X chainlocks: I was thinking about it a bit and... we know that chainlock locks the active chain only and there can only be one active chain, so why keeping more than one chainlock? 🤔